packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,902 @@
+                new bytes(0)
+            );
+
+            // Verify withdrawal data
+            // Start by reading the first 2 bytes into header
+            uint txsDataPtr = ctx.txsDataPtr + 2;
+            // header: txType (1) | type (1)
+            uint header;
+            // packedData: tokenID (2) | amount (12) | feeTokenID (2) | fee (2)
+            uint packedData;
+            bytes20 dataHash;
+            assembly {
+                header     := calldataload(    txsDataPtr     )
+                packedData := calldataload(add(txsDataPtr, 42))
+                dataHash   := and(calldataload(add(txsDataPtr, 78)), 0xffffffffffffffffffffffffffffffffffffffff000000000000000000000000)
+            }
+            require(
+                // txType == ExchangeData.TransactionType.WITHDRAWAL &&
+                // withdrawal.type == 1 &&
+                header & 0xffff == (uint(ExchangeData.TransactionType.WITHDRAWAL) << 8) | 1 &&
+                // withdrawal.tokenID == token.tokenID &&
+                // withdrawal.amount == token.amount &&
+                // withdrawal.fee == 0,
+                packedData & 0xffffffffffffffffffffffffffff0000ffff == (uint(token.tokenID) << 128) | (token.amount << 32) &&
+                onchainDataHash == dataHash,
+                "INVALID_BRIDGE_WITHDRAWAL_TX_DATA"
+            );
+
+            ctx.txsDataPtr += ExchangeData.TX_DATA_AVAILABILITY_SIZE;
+        }
+    }
+
+    function _storeTransfers(
+        bytes memory transfers,
+        address from
+        )
+        internal
+    {
+        uint batchID = batchIDGenerator++;
+
+        // Store transfers to distribute at a later time
+        bytes32 hash = _hashTransferBatch(transfers);
+        require(pendingTransfers[batchID][hash] == 0, "DUPLICATE_BATCH");
+        pendingTransfers[batchID][hash] = block.timestamp;
+
+        // Log transfers to do
+        emit Transfers(batchID, transfers, from);
+    }
+
+    function _deposit(
+        address from,
+        address token,
+        uint96  amount
+        )
+        internal
+    {
+        if (amount == 0) {
+            return;
+        }
+
+        if (from == address(this) && token != address(0)) {
+            ERC20(token).approve(address(depositContract), amount);
+        }
+        // Do the token transfer directly to the exchange
+        uint ethValue = (token == address(0)) ? amount : 0;
+        exchange.deposit{value: ethValue}(from, address(this), token, amount, new bytes(0));
+    }
+
+    function _connectorCall(
+        Context          memory   ctx,
+        ConnectorCalls   calldata connectorCalls,
+        uint                      n,
+        ConnectorCalls[] calldata allConnectorCalls
+        )
+        internal
+        returns (BridgeTransfer[] memory transfers)
+    {
+        require(connectorCalls.connector != address(this), "INVALID_CONNECTOR");
+        require(trustedConnectors[connectorCalls.connector], "ONLY_TRUSTED_CONNECTORS_SUPPORTED");
```

Comment:
So untrusted connectors are not yet supported? I remember I saw code that support untrusted ones, did you remove it?

Reply:
Yes, I removed it for now because trusted connectors allow for more optimizations:
- Tokens don't have to be transferred to the connector to keep things secure
- All returned bridge transfers from trusted connectors can be grouped together in a single batch deposit. For untrusted connectors this is not possible because they can return garbage data that could make all connectors or even the complete Ethereum transaction fail.

I'm assuming we will (definitely in the short and medium term) will only use trusted connector even just for UI reasons because we want to know what actually happens. And of course also because it's cheaper. So didn't think the extra complexity was worth it for now.